### PR TITLE
feat: inline constant tag transformation

### LINKED ISSUE #115: inline constant tag name in JSX transformation
### Description

Constant tag name should be inlined into the JSX tag to reduce the noise.

source: https://github.com/pionxzh/wakaru/issues/113#issuecomment-1946779728

### Input code

```typescript
const Name = "div";
const attrs = {id: "x"};
const Comp = React.createElement(Name, attrs);
```

### Expected behavior

```typescript
const attrs = {id: "x"};
const Comp = <div {...attrs} />;
```

### Actual behavior

```jsx
const Name = "div";
const attrs = { id: "x" };
const Comp = <Name {...attrs} />;
```

### PR DESCRIPTION
**Pull request by AhmadFaraz-crypto**
Closes #115

### CODE REVIEW

**Comment by pionxzh** [comment]
hm you might need to rebase to get latest changes

**Comment by pionxzh** [comment]
and please add corresponding test cases. Let me know if you need some help on this.

**Comment by AhmadFaraz-crypto** [comment]
> and please add corresponding test cases. Let me know if you need some help on this.

Sure, Thank you!

**Comment by AhmadFaraz-crypto** [comment]
Sorry about so many commits, I didn't check the test case before.

**Comment by pionxzh** [comment]
hmm let me figure out how to rebase your branch

**Comment by AhmadFaraz-crypto** [comment]
I just rebase my branch.

**Comment by pionxzh** [comment]
I will review this pr tomorror 🙏

File: packages/unminify/src/transformations/__tests__/un-jsx.spec.ts
Comment (j4k0xb) [inline packages/unminify/src/transformations/__tests__/un-jsx.spec.ts:166]: why is this inlined? `r` isn't constant

Comment (pionxzh) [reply packages/unminify/src/transformations/__tests__/un-jsx.spec.ts:166]: yes, this is not correct

Comment (AhmadFaraz-crypto) [reply packages/unminify/src/transformations/__tests__/un-jsx.spec.ts]: I just update the test case.

File: packages/unminify/src/transformations/__tests__/un-jsx.spec.ts
Comment (pionxzh) [inline packages/unminify/src/transformations/__tests__/un-jsx.spec.ts:157]: This test case shouldn't be changed. This is for "dynamic" Component tags.

File: packages/unminify/src/transformations/__tests__/un-jsx.spec.ts
Comment (pionxzh) [inline packages/unminify/src/transformations/__tests__/un-jsx.spec.ts:174]: Same here.

File: packages/unminify/src/transformations/un-jsx.ts
Comment (pionxzh) [inline packages/unminify/src/transformations/un-jsx.ts:135]: can use `j.JSXIdentifier.check` to narrow the type.

File: packages/unminify/src/transformations/__tests__/un-jsx.spec.ts
Comment (AhmadFaraz-crypto) [inline packages/unminify/src/transformations/__tests__/un-jsx.spec.ts:212]: I think I have added the dynamic components test case.

File: packages/unminify/src/transformations/__tests__/un-jsx.spec.ts
Comment (AhmadFaraz-crypto) [reply packages/unminify/src/transformations/__tests__/un-jsx.spec.ts:157]: If we are talking about, I should initialise the `r` then I think I have added this test case -> https://github.com/pionxzh/wakaru/pull/129/files#r1590658669

File: packages/unminify/src/transformations/__tests__/un-jsx.spec.ts
Comment (AhmadFaraz-crypto) [reply packages/unminify/src/transformations/__tests__/un-jsx.spec.ts:174]: previously it was working like 
input 
```
React.createElement(components[0], null, "Hello");
```
output 
```
const Component = components[0];
<Component>Hello</Component>; 
```

It was creating a variable as a tag and look like this was not picking components[0] also there were not intialized components variable.


According to this PR flow, I can update this, please check below example 

input

```
const Components = ['div']
React.createElement(
    Components[0],
    null,
    "Hello",
  );
```

output 

```
const Components = ['div'];
<div>Hello</div>
```

Comment (pionxzh) [reply packages/unminify/src/transformations/__tests__/un-jsx.spec.ts:174]: Oh, then it's ok. My suggestion is that we should always keep existing test cases. We can update the output according to the changes and add some new test cases to demonstrate the new behavior.

---

So we are going to do some evaluation things to determine if we can inline the tag?

File: packages/unminify/src/transformations/__tests__/un-jsx.spec.ts
Comment (pionxzh) [inline packages/unminify/src/transformations/__tests__/un-jsx.spec.ts:202]: the script has no idea about the value of `p` and `r`. This one should not be inlined.

File: packages/unminify/src/transformations/__tests__/un-jsx.spec.ts
Comment (AhmadFaraz-crypto) [reply packages/unminify/src/transformations/__tests__/un-jsx.spec.ts:174]: Noted. Yep, I will update the MR description, as I have handled a few scenarios with this MR and the ticket only mentions one. And let me update the code for the above test case.

File: packages/unminify/src/transformations/__tests__/un-jsx.spec.ts
Comment (pionxzh) [reply packages/unminify/src/transformations/__tests__/un-jsx.spec.ts:212]: This test is ok. But the value and `const Component` should be removed. I know it would be too complex to implement, you can leave it to me 😄

Comment (AhmadFaraz-crypto) [reply packages/unminify/src/transformations/__tests__/un-jsx.spec.ts:212]: haha, I can do it a try to remove this.

Comment (pionxzh) [reply packages/unminify/src/transformations/__tests__/un-jsx.spec.ts:212]: check `removeDeclarationIfUnused()`

**Comment by pionxzh** [comment]
Hi, I just reviewed and squashed all the commits. Please jump to the latest before you add any commits. 🙏 

I have to say sorry that I removed the ternary and array inlining. Here are the reasons:
1. The current Implementation will fail on a lot of edge cases. You can jump to the commit that I added test cases to see what's wrong.
2. Constant folding or let's say the "compression" in SWC/Terser is difficult. We need to do a lot of work to make it "correct". If we want to add this ability to `wakaru`, then it should be added as a standalone rule that inlines variables and constant results. It's not a JSX-specific thing.
3. Most modern compilers alr do the compression for you. So we can confidently say these are rare cases.

**Comment by 0xdevalias** [comment]
> I just reviewed and squashed all the commits

@pionxzh As an aside, a better workflow for future may be [squash commits when merging](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/about-pull-request-merges#squash-and-merge-your-commits), rather than squashing  a contributors working branch. I can't speak for all, but if a maintainer squashed and force pushed the branch I was working on, that could be enough to make me just abandon the PR all together.

**Comment by pionxzh** [comment]
@0xdevalias Thanks for the reminder. That's definitely not a good workflow.

@AhmadFaraz-crypto Sorry for what I did. This PR was stuck at a point where there were conflicts with the main branch. I was trying to resolve the conflicts, and I shouldn't squash the commits in the process. I will make sure to not do it again.

I'm away from my PC right now, but I want to talk more about why some of the changes were made. I will do that as soon as I get back.

**Comment by pionxzh** [comment]
#115 still needs the props inlining to be considered "completed". But let's merge this first.

**Review by pionxzh — APPROVED** [review]

**Pull request merged**
